### ISSUE TITLE
Alias cannot refer to other aliases.

### ISSUE THREAD

**Issue by SmartViking** [scopatz/xonsh#109]
```
uglemat@uglemat ~ $ aliases[".."] = ["cd", ".."]
uglemat@uglemat ~ $ ..
xonsh: subprocess mode: command not found: cd
Did you mean one of the following?
    dc: Command (/usr/bin/dc)
    cd: Alias
    cc: Command (/usr/bin/cc)
    dd: Command (/bin/dd)
    mcd: Command (/usr/bin/mcd)
uglemat@uglemat ~ $ def up(args, stdin=None):
...................   $[cd ..]
...................   return
................... 
uglemat@uglemat ~ $ aliases[".."] = up
uglemat@uglemat ~ $ up
<function up at 0x7f34fba37158>
uglemat@uglemat ~ $ ..
uglemat@uglemat /home $
 ```

I think the problem is that "cd" is another alias. You could argue that it wouldn't be a very good behavior for alias to recursively evaluate the first token, but `cd` is such a common command that the above should be possible. So the two obvious solutions seem to be either

1. recursive evaluation (give a thought to infinite loops, it might make sense to let it happen, or maybe they should be prevented by a "seen-tokens" set)
2. don't make very common commands like `cd` aliases

**Comment by scopatz** [comment]
Interesting point. This should be supported!

**Comment by SmartViking** [comment]
I can implement the recursive evaluation (planning to do it in `built_ins.Aliases`. I need to change the interface a bit because `run_subproc` needs the extra "expanded" args when the alias is a `callable` (the base case)), but I need to know what the expected behavior/purpose is for string aliases. This is the current behavior:
```
uglemat@uglemat ~/Public/xonsh master $ aliases['lal'] = '["ls", "-a"]'
uglemat@uglemat ~/Public/xonsh master $ lal
.   .git	.travis.yml  build  license	parser_test_table.py  scripts	tests  xonsh.egg-info
..  .gitignore	README.rst   docs   parser.out	recipe		      setup.py	xonsh
uglemat@uglemat ~/Public/xonsh master $ aliases['hay'] = 'print("HAY")'
uglemat@uglemat ~/Public/xonsh master $ hay
HAY
xonsh: subprocess mode: command not found: hay
Did you mean one of the following?
    hay:  Alias
    say:  Command (/usr/bin/say)
    ar:   Command (/usr/bin/ar)
    la:   Alias
    as:   Command (/usr/bin/as)
uglemat@uglemat ~/Public/xonsh master $ 
```
Is this correct behavior? Can I rely on the eval'd string-value to return a list, like with `lal`?

**Comment by scopatz** [comment]
>  Can I rely on the eval'd string-value to return a list, like with lal?

Probably not.  The intended behaviour was to have a anonymous block of code that is executed.  Unlike with function aliases, string aliases don't take any input and they don't return any output, per say.  It is up to the alias author to determine if the return value of `evalx()` is valid input to the next thing in the pipeline.

**Comment by scopatz** [comment]
Also @SmartViking - thanks a ton for taking a stab at this!

**Comment by SmartViking** [comment]
I pretty much have this implemented now, but not the "string alias". 

```
uglemat@uglemat ~/Public/xonsh master $ aliases['hello']='ls -a'
uglemat@uglemat ~/Public/xonsh master $ hello
.     .gitignore   __pycache__  license               recipe    tests
..    .travis.yml  build        parser.out            scripts   xonsh
.git  README.rst   docs         parser_test_table.py  setup.py  xonsh.egg-info
xonsh: subprocess mode: command not found: hello
Did you mean the following?
    hello:  Alias
uglemat@uglemat ~/Public/xonsh master $ 
```
Isn't this a bug? It runs the code, but then it tries run "hello" afterwards. Would it be out of the question to remove that feature? I seems like a confusing element. Are there things you can do with string-alias that you can't do with callable-alias? Are there things which are more practical to do with string-alias than callable-alias? The lack of examples makes this string-alias kind of a mystery to me :)

**Comment by scopatz** [comment]
This is definitely a bug

**Comment by scopatz** [comment]
>Would it be out of the question to remove that feature? I seems like a confusing element. 

No not at all.

> Are there things you can do with string-alias that you can't do with callable-alias? 

nope.

> Are there things which are more practical to do with string-alias than callable-alias?

Well, the reason for string aliases was a convenience for list aliases.

> The lack of examples makes this string-alias kind of a mystery to me :)

Yeah, looking at the code now, it is a little confusing what their purpose should be.  Perhaps the right thing to do is for the `Aliases` class to automatically call `shlex.split()` on any string values that are added, prior to inserting them. What do you think about that, @SmartViking?

**Comment by SmartViking** [comment]
Sounds like a good idea, I'll do that.

**Comment by SmartViking** [comment]
I've implemented it in #115. I made some tests too for `Aliases`, because this evaluation stuff is so delicate. I haven't used nosetest before, so I just tried to imitate the other files.
